lib/eclipse/EclipseState/Schedule/MSW/Compsegs.cpp

```diff
@@ -168,18 +168,61 @@ namespace Opm {
             }
 
             compseg.m_segment_number = segment_number;
+
+            // when depth is default or zero, we obtain the depth of the completion based on the information
+            // of the related segments
             if (compseg.m_center_depth == 0.) {
-                // using the depth of the segment node as the depth of the completion
-                // TODO: now only one completion for one segment is hanlded,
-                // TODO: later we will try to handle more than one completion for each segment,
-                // TODO: which will be a linear interpolation based on the segment node depth
-                // TODO: in the same branch, while the actually way is not clear yet
-                const int segment_location = segment_set.numberToLocation(segment_number);
-                compseg.m_center_depth = segment_set[segment_location].depth();
+                compseg.calculateCenterDepthWithSegments(segment_set);
             }
         }
     }
 
+    void Compsegs::calculateCenterDepthWithSegments(const SegmentSet& segment_set) {
+
+        // the depth and distance of the segment to the well head
+        const int segment_location = segment_set.numberToLocation(m_segment_number);
+        const double segment_depth = segment_set[segment_location].depth();
+        const double segment_distance = segment_set[segment_location].totalLength();
+
+        // for top segment or the distance of segment and perforation are the same
+        // no interpolation is needed
+        const double center_distance = (m_distance_start + m_distance_end) / 2.0;
+        if (center_distance == segment_distance || m_segment_number == 1) {
+            m_center_depth = segment_depth;
+            return;
+        }
+
+        // for other cases, interpolation between two segments is needed.
+        // looking for the other segment needed for interpolation
+        // by default, it uses the outlet segment to do the interpolation
+        int interpolation_segment = segment_set[segment_location].outletSegment();
+
+        // if the perforation is further than the segment and the segment has inlet segments in the same branch
+        // we use the inlet segment to do the interpolation
+        if (center_distance > segment_distance) {
+            for (const int inlet : segment_set[segment_location].inletSegments()) {
+                const int inlet_location = segment_set.numberToLocation(inlet);
+                if (segment_set[inlet_location].branchNumber() == m_branch_number) {
+                    interpolation_segment = inlet;
+                    break;
+                }
+            }
+        }
+
+        assert(interpolation_segment != 0);
```

Comment:
This seems so fragile it should be protected in release as well - i.e.:
```C++
if (interpolation_segment == 0) throw std::runtime_error(" ... ")
```

Reply:
fixed.